File: src/main_node.cc
#include <iostream>
#include <math.h>

#include <ros/ros.h>
#include <tf/tf.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/Twist.h>

namespace Robot
{
    class Motion
    {
        private:
            ros::NodeHandle nh;
            
            ros::Publisher pub;
            ros::Subscriber sub;

            nav_msgs::Odometry robot_odom;
            geometry_msgs::Twist robot_velocity;

            double x,y,z,roll,pitch,yaw;

            double x_goal = 4.0;
            double y_goal = 4.0;

            double K_linear = 0.8;
            double K_angular = 1.2;

            float distance;
            float desired_angle;
            
            float linear_speed;
            float angular_speed;

        public:
            Motion()
            {
                pub = nh.advertise<geometry_msgs::Twist>("cmd_vel",1000);
                sub = nh.subscribe<nav_msgs::Odometry>("odometry/filtered",1000,&Motion::odom_callback,this);
            }

            void odom_callback(const nav_msgs::Odometry::ConstPtr& msg)
            {
                x = msg->pose.pose.position.x;
                y = msg->pose.pose.position.y;
                z = msg->pose.pose.position.z;

                tf::Quaternion q(
                    msg->pose.pose.orientation.x,
                    msg->pose.pose.orientation.y,
                    msg->pose.pose.orientation.z,
                    msg->pose.pose.orientation.w);

                tf::Matrix3x3 m(q);

                m.getRPY(roll, pitch, yaw);
                
                robot_move();
            }

            void robot_move()
            {
                distance = abs(pow(x_goal-x,2)+pow(y_goal-y,2));
                linear_speed = distance * K_linear;

                desired_angle = atan2(y_goal-y, x_goal-x);
                angular_speed = (desired_angle-yaw)*K_angular;

                robot_velocity.linear.x = linear_speed;
                robot_velocity.angular.z = angular_speed;

                if(distance<0.01)
                {
                    robot_velocity.linear.x=0.0;
                    robot_velocity.angular.z=0.0;
                    pub.publish(robot_velocity);
                    ros::shutdown();
                }

                ROS_INFO("X: %f Y: %f Theta: %f",x,y,yaw);

                pub.publish(robot_velocity);

            }
    };
}

int main(int argc, char** argv)
{
    ros::init(argc,argv,"main_node");

    Robot::Motion jackal;

    ros::spin();

    return 0;
}

File: scripts/main_node.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
import math
import time
import tf

x=0
y=0
z=0

roll=0
pitch=0
yaw=0

def poseCallback(odom_msg):
    
    global x,y,z
    global roll,pitch,yaw
    
    x = odom_msg.pose.pose.position.x
    y = odom_msg.pose.pose.position.y
    z = odom_msg.pose.pose.position.z

    quaternion = (
    odom_msg.pose.pose.orientation.x,
    odom_msg.pose.pose.orientation.y,
    odom_msg.pose.pose.orientation.z,
    odom_msg.pose.pose.orientation.w)
    
    #convert the quaternion to roll-pitch-yaw
    rpy = tf.transformations.euler_from_quaternion(quaternion)
    
    #extract the values of roll, pitch and yaw from the array
    roll = rpy[0]
    pitch = rpy[1]
    yaw = rpy[2]

    print("X: {} Y: {} Theta:{}".format(x,y,yaw))


def move(speed, distance, is_forward):
        #declare a Twist message to send velocity commands
        velocity_message = Twist()
        #get current location 
        global x, y
        x0=x
        y0=y

        if (is_forward):
            velocity_message.linear.x =abs(speed)
        else:
            velocity_message.linear.x =-abs(speed)

        distance_moved = 0.0
        loop_rate = rospy.Rate(10) # we publish the velocity at 10 Hz (10 times a second)    
        cmd_vel_topic='/turtle1/cmd_vel'
        velocity_publisher = rospy.Publisher(cmd_vel_topic, Twist, queue_size=10)

        while True :
                rospy.loginfo("Turtlesim moves forwards")
                velocity_publisher.publish(velocity_message)

                loop_rate.sleep()
                
                #rospy.Duration(1.0)
                
                distance_moved = abs(0.5 * math.sqrt(((x-x0) ** 2) + ((y-y0) ** 2)))
                print(distance_moved)               
                if  not (distance_moved<distance):
                    rospy.loginfo("reached")
                    break
        
        #finally, stop the robot when the distance is moved
        velocity_message.linear.x =0
        velocity_publisher.publish(velocity_message)
    
def rotate (angular_speed_degree, relative_angle_degree, clockwise):
    
    global yaw
    velocity_message = Twist()
    velocity_message.linear.x=0
    velocity_message.linear.y=0
    velocity_message.linear.z=0
    velocity_message.angular.x=0
    velocity_message.angular.y=0
    velocity_message.angular.z=0

    #get current location 
    theta0=yaw
    angular_speed=math.radians(abs(angular_speed_degree))

    if (clockwise):
        velocity_message.angular.z =-abs(angular_speed)
    else:
        velocity_message.angular.z =abs(angular_speed)

    angle_moved = 0.0
    loop_rate = rospy.Rate(10) # we publish the velocity at 10 Hz (10 times a second)    
    cmd_vel_topic='/turtle1/cmd_vel'
    velocity_publisher = rospy.Publisher(cmd_vel_topic, Twist, queue_size=10)

    t0 = rospy.Time.now().to_sec()

    while True :
        rospy.loginfo("Turtlesim rotates")
        velocity_publisher.publish(velocity_message)

        t1 = rospy.Time.now().to_sec()
        current_angle_degree = (t1-t0)*angular_speed_degree
        loop_rate.sleep()


                       
        if  (current_angle_degree>relative_angle_degree):
            rospy.loginfo("reached")
            break

    #finally, stop the robot when the distance is moved
    velocity_message.angular.z =0
    velocity_publisher.publish(velocity_message)


def go_to_goal(x_goal, y_goal):

    velocity_message = Twist()

    while (True):
        K_linear = 0.5 
        distance = abs(math.sqrt(((x_goal-x) ** 2) + ((y_goal-y) ** 2)))

        linear_speed = distance * K_linear


        K_angular = 4.0
        desired_angle_goal = math.atan2(y_goal-y, x_goal-x)
        angular_speed = (desired_angle_goal-yaw)*K_angular

        velocity_message.linear.x = linear_speed
        velocity_message.angular.z = angular_speed

        velocity_publisher.publish(velocity_message)

        print('x=', x, 'y=',y,'theta',yaw)


        if (distance <0.01):
            break

def setDesiredOrientation(desired_angle_radians):
    relative_angle_radians = desired_angle_radians - yaw
    if relative_angle_radians < 0:
        clockwise = 1
    else:
        clockwise = 0
    print(relative_angle_radians)
    print(desired_angle_radians)
    rotate(30 ,math.degrees(abs(relative_angle_radians)), clockwise)

def gridClean():
 
    desired_pose = Pose()
    desired_pose.x = 1
    desired_pose.y = 1
    desired_pose.theta = 0
 
    moveGoal(desired_pose, 0.01)
 
    setDesiredOrientation(degrees2radians(desired_pose.theta))
 
    move(2.0, 9.0, True)
    rotate(degrees2radians(20), degrees2radians(90), False)
    move(2.0, 9.0, True)
    rotate(degrees2radians(20), degrees2radians(90), False)
    move(2.0, 1.0, True)
    rotate(degrees2radians(20), degrees2radians(90), False)
    move(2.0, 9.0, True)
    rotate(degrees2radians(30), degrees2radians(90), True)
    move(2.0, 1.0, True)
    rotate(degrees2radians(30), degrees2radians(90), True)
    move(2.0, 9.0, True)
    pass
 
 
def spiralClean():
    vel_msg = Twist()
    loop_rate = rospy.Rate(1)
    wk = 4
    rk = 0
 
    while((currentTurtlesimPose.x<10.5) and (currentTurtlesimPose.y<10.5)):
        rk=rk+1
        vel_msg.linear.x =rk
        vel_msg.linear.y =0
        vel_msg.linear.z =0
        vel_msg.angular.x = 0
        vel_msg.angular.y = 0
        vel_msg.angular.z =wk
        velocity_publisher.publish(vel_msg)
        loop_rate.sleep()
 
    vel_msg.linear.x = 0
    vel_msg.angular.z = 0
    velocity_publisher.publish(vel_msg)



if __name__ == '__main__':
    try:
        
        rospy.init_node('jackal_motion', anonymous=True)

        # cmd_vel publisher
        cmd_vel_topic='cmd_vel'
        velocity_publisher = rospy.Publisher(cmd_vel_topic, Twist, queue_size=10)
        
        position_topic = "odometry/filtered"
        pose_subscriber = rospy.Subscriber(position_topic, Odometry, poseCallback) 
        
        time.sleep(2)

        #move(1.0, 2.0, False)
        #rotate(30, 90, True)
        go_to_goal(4.0,4.0)
        #setDesiredOrientation(math.radians(90))
       
    except rospy.ROSInterruptException:
        rospy.loginfo("node terminated.")